### ISSUE TITLE
Virtual Downloads not generated on the fly

### ISSUE THREAD

**Issue by dathwa** [OFFLINE-GmbH/oc-mall-plugin#878]
Hi. 
Regarding this update: #862, I believe this change has broken how downloads are dynamically generated.
My PDF downloads are generated on the fly, so there is never any file attached to these products.
Is this intentional? 
Thank-you.

I had to comment out these lines...
```
public static function fromOrderProduct(OrderProduct $orderProduct)
    {
        
        // Do not create a file grant if the product does not have a file associated to it
        // if (!$orderProduct->product->latest_file) {
        //     return;
        // }
        
...
    
    }
```

**Comment by tobias-kuendig** [comment]
Thank you for bringing this up. It seems like the newly added lines prevent the Event from being fired. We probably need to fire the event in any case for each product, but not create the grant.

**Comment by dathwa** [comment]
> Thank you for bringing this up. It seems like the newly added lines prevent the Event from being fired. We probably need to fire the event in any case for each product, but not create the grant.

Agreed. Also, for some of my virtual products (bookings), I need the 'mall.checkout.succeeded' event fired, when I create a booking for them to access in their account. But i don't require a ticket (file). So, I don't want to attach a fixed file to the product, and i don't want the attach a dynamic file to the product. This would ideally remove the link from the customer order in frontend, and not trigger the downloads email to the customer. Would this scenario be possible? Thank-you.

**Comment by tobias-kuendig** [comment]
This has been fixed in b7414a2, thank you for this report!
